src/gap.c

```diff
@@ -1374,75 +1341,91 @@ Obj FuncSetUserHasQuit( Obj Self, Obj value)
    
  Obj FuncCALL_WITH_TIMEOUT( Obj self, Obj seconds, Obj microseconds, Obj func, Obj args )
 {
-  Obj plain_args;
     Obj res;
     Obj currLVars;
     Obj result;
     Int recursionDepth;
     Stat currStat;
+    UInt newJumpBuf = 1;
+    UInt mayNeedToRestore = 0;
+    Int iseconds, imicroseconds;
+    Int curr_seconds, curr_microseconds, curr_nanoseconds;
+    Int restore_seconds = 0, restore_microseconds = 0;
+      
     if (!SyHaveAlarms)
       ErrorMayQuit("CALL_WITH_TIMEOUT: timeouts not supported on this system", 0L, 0L);
     if (!IS_INTOBJ(seconds) || 0 > INT_INTOBJ(seconds))
-      ErrorMayQuit("CALL_WITH_TIMEOUT(<seconds>, <microseconds>, <func>,<args>): <seconds> must be a non-negative small integer",0,0);
+      ErrorMayQuit("CALL_WITH_TIMEOUT(<seconds>, <microseconds>, <func>,<args>):"
+		   " <seconds> must be a non-negative small integer",0,0);
     if (!IS_INTOBJ(microseconds) || 0 > INT_INTOBJ(microseconds) || 999999999 < INT_INTOBJ(microseconds))
-      ErrorMayQuit("CALL_WITH_TIMEOUT(<seconds>, <microseconds>, <func>,<args>): <microseconds> must be a non-negative small integer less than 10^9",0,0);
+      ErrorMayQuit("CALL_WITH_TIMEOUT(<seconds>, <microseconds>, <func>,<args>):"
+		   " <microseconds> must be a non-negative small integer less than 10^9",0,0);
+    iseconds = INT_INTOBJ(seconds);
+    imicroseconds = INT_INTOBJ(microseconds);
     if (!IS_FUNC(func))
       ErrorMayQuit("CALL_WITH_TIMEOUT(<seconds>, <microseconds>, <func>,<args>): <func> must be a function",0,0);
     if (!IS_LIST(args))
       ErrorMayQuit("CALL_WITH_TIMEOUT(<seconds>, <microseconds>, <func>,<args>): <args> must be a list",0,0);
-    if (!IS_PLIST(args))
-      {
-        plain_args = SHALLOW_COPY_OBJ(args);
-        PLAIN_LIST(plain_args);
+    if (SyAlarmRunning) {            
+      /* ErrorMayQuit("CALL_WITH_TIMEOUT cannot currently be nested except via break loops."
+	 " There is already a timeout running", 0, 0); */
+      SyStopAlarm((UInt *)&curr_seconds, (UInt *)&curr_nanoseconds);
+      curr_microseconds = curr_nanoseconds/1000;
+      if (iseconds > curr_seconds || (iseconds == curr_seconds && imicroseconds > curr_microseconds)) {
+	  /* The existing timeout will go off before we would so don't bother to set a timeout
+	   just reset the existing one and call the function */
+	  iseconds = curr_seconds;
+	  imicroseconds = curr_microseconds;
+	  newJumpBuf = 0;
+	} else {
+	/* We would time out before the other function, so we need to
+	   set our timeout, but remember to reset things later */
+	mayNeedToRestore = 1;
+	newJumpBuf = 1;
```

Comment:
Isn't the indention level for this last `else` block wrong again? It looks like that in the github interface, at least...